File: leetcode/C++/max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int sol=0,t=0;
        for(int l=0,r=0; r<nums.size(); ){
            if(nums[r]==1) {++t; ++r;}
            else if(k>0) {--k; ++t; nums[r++]=-1;}
            else if(nums[l]==0) ++l;
            else if(nums[l]==1) {++l; --t;}
            else {++l; --t; ++k;}
            if(l>=r) ++r;
            sol=max(sol,t);
        }
        return sol;
    }
};

File: leetcode/Python/max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        sol, t, l, r = 0, 0, 0, 0
        while r<len(nums):
            if nums[r]==1:
                t, r = t+1, r+1
            elif k>0:
                k,t,nums[r],r=k-1,t+1,-1,r+1
            elif nums[l]==0:
                l+=1
            elif nums[l]==1:
                l, t = l+1, t-1
            else:
                l,t,k = l+1, t-1, k+1
            if l>=r:
                r+=1
            sol=max(sol,t)
        return sol